cluster/juju/layers/kubernetes-master/reactive/kubernetes_master.py

```diff
@@ -1408,37 +1493,100 @@ def clear_cluster_tag_sent():
     remove_state('kubernetes-master.cluster-tag-sent')
 
 
-@when('endpoint.aws.joined',
-      'leadership.set.cluster_tag')
-@when_not('kubernetes-master.aws-request-sent')
+@when_any('endpoint.aws.joined',
+          'endpoint.gcp.joined')
+@when('leadership.set.cluster_tag')
+@when_not('kubernetes-master.cloud-request-sent')
 def request_integration():
-    hookenv.status_set('maintenance', 'requesting aws integration')
-    aws = endpoint_from_flag('endpoint.aws.joined')
+    hookenv.status_set('maintenance', 'requesting cloud integration')
     cluster_tag = leader_get('cluster_tag')
-    aws.tag_instance({
-        'KubernetesCluster': cluster_tag,
-        'k8s.io/role/master': 'true',
-    })
-    aws.tag_instance_security_group({
-        'KubernetesCluster': cluster_tag,
-    })
-    aws.enable_instance_inspection()
-    aws.enable_network_management()
-    aws.enable_dns_management()
-    aws.enable_load_balancer_management()
-    aws.enable_block_storage_management()
-    aws.enable_object_storage_management(['kubernetes-*'])
-    set_state('kubernetes-master.aws-request-sent')
-    hookenv.status_set('waiting', 'waiting for aws integration')
-
-
-@when_not('endpoint.aws.joined')
+    if is_state('endpoint.aws.joined'):
+        cloud = endpoint_from_flag('endpoint.aws.joined')
+        cloud.tag_instance({
+            'kubernetes.io/cluster/{}'.format(cluster_tag): 'owned',
+            'k8s.io/role/master': 'true',
+        })
+        cloud.tag_instance_security_group({
+            'kubernetes.io/cluster/{}'.format(cluster_tag): 'owned',
+        })
+        cloud.tag_instance_subnet({
+            'kubernetes.io/cluster/{}'.format(cluster_tag): 'owned',
+        })
+        cloud.enable_object_storage_management(['kubernetes-*'])
+        cloud.enable_load_balancer_management()
+    elif is_state('endpoint.gcp.joined'):
+        cloud = endpoint_from_flag('endpoint.gcp.joined')
+        cloud.label_instance({
+            'k8s-io-cluster-name': cluster_tag,
+            'k8s-io-role-master': 'master',
+        })
+        cloud.enable_object_storage_management()
+        cloud.enable_security_management()
+    cloud.enable_instance_inspection()
+    cloud.enable_network_management()
+    cloud.enable_dns_management()
+    cloud.enable_block_storage_management()
+    set_state('kubernetes-master.cloud-request-sent')
+    hookenv.status_set('waiting', 'waiting for cloud integration')
```

Comment:
This status message will need to go in [set_final_status](https://github.com/kubernetes/kubernetes/blob/dbc491c03102ca882e8d1b4a00c79cf36ff903fe/cluster/juju/layers/kubernetes-master/reactive/kubernetes_master.py#L393) if you don't want it to get clobbered by other messages.

Reply:
Oops. I missed that, and I'm not okay with it. Please remove `layer:status`. Every new layer we depend on increases our maintenance burden. See for example: https://github.com/juju-solutions/bundle-canonical-kubernetes/issues/552 where a change in layer-snap broke us. If we don't need a dependency, we should not use it.

Reply:
I disagree, this layer is purely operational, whereas the snap layer is functional and breakages are problematic. This just streamlining status messages should be a pretty minimal level of impact and wont' be one that breaks arbitrarily

Reply:
@marcoceppi The status layer doesn't really add anything to these charms, though, because of the single `set_final_status` handler.  I do think it will be useful for other charms to not need such a handler and to be able to manage status in a more natural way, but these charms are already pretty big and it's reasonable to limit introducing new layers at this point.